File: LeetCodePractice/src/DP/New21Game.java
package DP;
/*Alice plays the following game, loosely based on the card game "21".

Alice starts with 0 points, and draws numbers while she has less than K points.  During each draw, she gains an integer number of points randomly from the range [1, W], where W is an integer.  Each draw is independent and the outcomes have equal probabilities.

Alice stops drawing numbers when she gets K or more points.  What is the probability that she has N or less points?

Example 1:

Input: N = 10, K = 1, W = 10
Output: 1.00000
Explanation:  Alice gets a single card, then stops.
Example 2:

Input: N = 6, K = 1, W = 10
Output: 0.60000
Explanation:  Alice gets a single card, then stops.
In 6 out of W = 10 possibilities, she is at or below N = 6 points.
Example 3:

Input: N = 21, K = 17, W = 10
Output: 0.73278
Note:

0 <= K <= N <= 10000
1 <= W <= 10000
Answers will be accepted as correct if they are within 10^-5 of the correct answer.
The judging time limit has been reduced for this question.*/
public class New21Game {
	public double new21Game(int N, int K, int W) {
		if(K == 0 || K + W <= N){
			return 1.0;
		}
		int max = K + W - 1;
        double[] sum = new double[max + 1];
        // mem[x] means probability of get x point
        // mem[i] = (mem[i - W] + mem[i - W + 1] + ... + mem[i - 1]) / W
        // could be simplified to mem[i] = (sum[i - 1] - sum[i - W - 1]) / W.
        //  if we use sum[i], we can get sum[i] = sum[i - 1] + (sum[i - 1] - sum[i - W - 1]) / W.
        sum[0] = 1;
        // s = mem[K + 1] + mem[K + 2] + .... + mem[K + W];
        for(int i = 1; i <= max; i++){
        	if(i <= W){
        		sum[i] = sum[i - 1] + sum[i - 1] / W;
        	} else {
        		sum[i] = sum[i - 1] + (sum[i - 1] - sum[i - W - 1]) / W;
        	}
        	if(i > K){
        		sum[i] -= (sum[i - 1] - sum[K - 1]) / W;
        	}
        }
        return sum[N] - sum[K - 1];
    }
}

File: LeetCodePython/New21Game.py
"""
Alice plays the following game, loosely based on the card game "21".
Alice starts with 0 points and draws numbers while she has less than k points. During each draw, she gains an integer number of points randomly from the range [1, maxPts], where maxPts is an integer.
Each draw is independent and the outcomes have equal probabilities.
Alice stops drawing numbers when she gets k or more points.
Return the probability that Alice has n or fewer points.
Answers within 10-5 of the actual answer are considered accepted.

Example 1:
Input: n = 10, k = 1, maxPts = 10
Output: 1.00000
Explanation: Alice gets a single card, then stops.

Example 2:
Input: n = 6, k = 1, maxPts = 10
Output: 0.60000
Explanation: Alice gets a single card, then stops.
In 6 out of 10 possibilities, she is at or below 6 points.

Example 3:
Input: n = 21, k = 17, maxPts = 10
Output: 0.73278

Constraints:
0 <= k <= n <= 10^4
1 <= maxPts <= 10^4

analysis:
dp[i] means the probability to get i points
logic:
probability to get i points = sum(probability to get i - x points * 1 / maxPts) for x can be any of [1, maxPts]
uses pre_sum_i_w to represent
where 0 <= i - w < K
"""

class New21Game:
    def new21Game(self, n: int, k: int, maxPts: int) -> float:
        if k == 0 or n >= k + maxPts:
            return 1
        dp = [0] * (n + 1)
        pre_sum_x, res = 1, 0
        dp[0] = 1
        for i in range(1, n + 1):
            dp[i] = pre_sum_x / maxPts
            if i < k:
                pre_sum_x += dp[i]
            else:
                res += dp[i]
            if i - maxPts >= 0:
                pre_sum_x -= dp[i - maxPts]
        return res
